src/st/dre/22848.c

```diff
@@ -176,11 +173,113 @@ void Entity3DHouseSpawner(Entity* self) {
             CreateEntityFromCurrentEntity(E_ID_BACKGROUND_HOUSE, newEntity);
             newEntity->posX.i.hi = *dataPtr++ - g_Tilemap.scrollX.i.hi;
             newEntity->posY.i.hi = *dataPtr++ - g_Tilemap.scrollY.i.hi;
+            // Params is either 0 or 1. In-game, there are 3 houses in
+            // a row, and behind them 2 more houses, which are rotated 90 deg.
+            // This likely controls the rotation.
             newEntity->params = *dataPtr++;
         }
         self->step++;
     }
 }
 
 // 3D house object in background ID 0x16
-INCLUDE_ASM("st/dre/nonmatchings/22848", Entity3DBackgroundHouse);
+void Entity3DBackgroundHouse(Entity* self) {
+    SVECTOR rot;
+    VECTOR trans;
+    MATRIX m;
+    s32 unused1;
+    s32 unused2;
+    Primitive* prim;
+    s32 primIndex;
+    s16* houseData;
+    s32 i;
+    SVECTOR** vertices;
+    s32* scratchpad;
+    u8* model;
+
+    switch (self->step) {
+    case 0:
+        InitializeEntity(g_EInitGeneric);
+        primIndex = g_api.AllocPrimitives(PRIM_GT4, 16);
+        // Usually this can be if primIndex == -1 DestroyEntity but nope
+        if (primIndex != -1) {
+            self->flags |= FLAG_HAS_PRIMS;
+            self->primIndex = primIndex;
+            prim = &g_PrimBuf[primIndex];
+            self->ext.bghouse.prim = prim;
+            for (; prim != NULL; prim = prim->next) {
+                prim->tpage = 0xF;
+                prim->priority = 0x58 - self->params;
+                prim->drawMode = DRAW_HIDE;
+            }
+
+        } else {
+            DestroyEntity(self);
+            return;
+        }
+        // These seem to be unused?
+        self->ext.bghouse.unk80 = 0x80;
+        self->ext.bghouse.unk82 = 0x80;
+        self->ext.bghouse.unk84 = 0x80;
+        return;
+    case 1:
+        // surprised this isn't just if abs(posX) > 0x200
+        if (self->posX.i.hi > 0x200 || self->posX.i.hi < -0x200) {
+            for (prim = self->ext.bghouse.prim; prim != NULL;
+                 prim = prim->next) {
+                prim->drawMode = 8;
+            }
+            return;
+        }
+        prim = self->ext.bghouse.prim; // unused
+        SetGeomScreen(0x400);
+        SetGeomOffset(128, 192);
+        // this is a 7-by-X array, so params picks a set of 7 values
+        houseData = &D_801815CC;
+        // Params is 0 or 1. 0 if house is straight, 1 if rotated 90 degree.
+        houseData += 7 * self->params;
+        rot.vx = 0;
+        rot.vy = *houseData++;
+        rot.vz = 0;
+        RotMatrix(&rot, &m);
+        trans.vx = self->posX.i.hi - 0x80;
+        trans.vy = self->posY.i.hi - 0xC0;
+        trans.vz = *houseData++ + 0x400;
+        TransMatrix(&m, &trans);
+        SetRotMatrix(&m);
+        SetTransMatrix(&m);
+        vertices = &D_80181544[0][0];
```

Comment:
```suggestion
        vertices = D_80181544;
```
can we remove the [0][0]?

Reply:
That came from PSP stuff, the PSP compiler is more aggressive about types matching, and it doesn't like doing `myPointer = myArray;` because the pointer is supposed to be a pointer to an int, not a pointer to an array of ints. I'll clean these up.

Reply:
It seems like we should keep the PSP match? We're going to combine PSP+PSX eventually

Reply:
Agree. If only the match `vertices = &D_80181544[0][0];` is valid on PSP, then we should keep it. Sorry for making you going back and forth on this.

Reply:
Sounds good.

It's also possible that there might be a compiler flag that tells it to not make errors on this kind of pointer type mismatch. In that case we could use the simpler expression for PSP, without needing &[0][0].